system/inc/asset_manager_api.h

```diff
@@ -0,0 +1,292 @@
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <stdint.h>
+#include <stddef.h>
+
+#ifdef __cplusplus
+
+#include "spark_wiring_string.h"
+#include "buffer.h"
+#include "bytes2hexbuf.h"
+
+namespace particle {
+
+// Shared hash class between application and system
+class AssetHash {
+public:
+    enum Type {
+        INVALID = -1,
+        SHA256 = 0,
+        DEFAULT = SHA256,
+    };
+
+    static const size_t SHA256_HASH_SIZE = 32;
+
+    AssetHash();
+    AssetHash(const char* hash, size_t length, Type type = Type::DEFAULT);
+    AssetHash(const uint8_t* hash, size_t length, Type type = Type::DEFAULT);
+    AssetHash(const Buffer& hash, Type type = Type::DEFAULT);
+    AssetHash(const AssetHash& other) = default;
+    AssetHash(AssetHash&& other) = default;
+
+    AssetHash& operator=(const AssetHash& other) = default;
+    AssetHash& operator=(AssetHash&& other) = default;
+
+    Type type() const;
+    const Buffer& hash() const;
+    bool isValid() const;
+
+    String toString() const;
+
+    bool operator==(const AssetHash& other) const;
+    bool operator!=(const AssetHash& other) const;
+
+private:
+    Type type_;
+    Buffer hash_;
+};
+
+inline AssetHash::AssetHash()
+        : type_(Type::INVALID) {
+}
+
+inline AssetHash::AssetHash(const char* hash, size_t length, Type type)
+        : AssetHash(Buffer(hash, length), type) {
+}
+
+inline AssetHash::AssetHash(const uint8_t* hash, size_t length, Type type)
+        : AssetHash(Buffer(hash, length), type) {
+}
+
+inline AssetHash::AssetHash(const Buffer& hash, Type type)
+        : AssetHash() {
+    // Only SHA256 for now
+    if (type == Type::SHA256 && hash.size() == SHA256_HASH_SIZE) {
+        hash_ = hash;
+        type_ = type;
+    }
+}
+
+inline AssetHash::Type AssetHash::type() const {
+    return type_;
+}
+
+inline const Buffer& AssetHash::hash() const {
+    return hash_;
+}
+
+inline bool AssetHash::isValid() const {
+    return type_ != Type::INVALID && hash_.size() > 0;
+}
+
+inline String AssetHash::toString() const {
+    if (hash_.size() > 0) {
+        Buffer buf(hash_.size() * 2 + 1);
+        if (buf.size()) {
+            memset(buf.data(), 0, buf.size());
+            bytes2hexbuf_lower_case((const uint8_t*)hash_.data(), hash_.size(), buf.data());
+            return String(buf.data());
+        }
+    }
+    return String();
+}
+
+inline bool AssetHash::operator==(const AssetHash& other) const {
+    return (type_ == other.type_ && hash_ == other.hash_);
```

Comment:
Is it intentional that the asset name is not taken into account here? It's worth adding a comment if so.